### DISCUSSION TITLE
Equipment Maintenance Process in ADempiere

### DISCUSSION THREAD

**Discussion by rvergara** [adempiere/adempiere#3974] [General]
### **Equipment Maintenance Process in ADempiere**
A few weeks ago, a presentation on, OFB consulting and Yamel and Mario refined, [Maintenance process for ADempiere](https://wiki.adempiere.net/Sponsored_Development:_Maintenance_Management_System) was presented on Discord. Unfortunately I could not attend so I decided to write this discussion to present my views which are based on 15 years'experience as an maintenance engineer.

Maintenance is a whole business process for many industry verticals, namely, transportation, mining, utilities, etc. It is so important for these verticals that many "best of breed" solutions were developed to support them (e.g. Maximo, MP and many others) until ERP solutions understood the importance of this business process and developed full blown solutions within their functional scope. An example of this is PM from SAP, SAP did not treat the maintenance process as an appendix of manufacturing (because it was not) but developed a whole new process to properly support those verticals mentioned above.

It is true that for some other verticals maintenace management is not a major requirement (e.g manufacturing) and a simple work order scheme would suffice.

However for those verticals mentioned above a proper full modern maintenance business process is required (verticals, that normally do not require manufacturing functionality). Such a process should include well determined funcionality for the major roles interacting in the process including maintenance planners, maintenance schedulers, maintainers, equipment operators and end users.

It is my opinion that the maintenance process developed by OFB et all should be included in ADempiere Trunk (same as SAP included PM).

Last but not least I have to add that maintenance cost are indeed included in this development for both maintenance personnel and parts which are defined per every work order task.

I would like to hear your opinions on this matter

**Comment by e-Evolution** [comment]
Hello everyone!

As the project's main maintainer and technical team member, I share with you the necessary requirements for a contribution to be added as a standard feature of ADempiere.

Technical requirements.

1.- The code in which it is written must follow the best existing coding practices. In particular, it is not recommended to modify the core classes of ADempiere, only if it is a correction, it is better to use model validators to overload a certain functionality or change it if necessary.

2.- The code, variables, classes, and methods must be completely developed using the English language.

3.- The development of the new functionality must extend from the development branch, it must be compatible with the version of java and databases that ADempiere supports (Oracle and PostgreSQL) in the last version

4.- The contribution must be sent to the ADempiere project from a fork and a pull request that extends from the recent development branch. it to review and feedback

5.- The data model proposed by the new functionality must be correctly integrated with the existing data model avoiding at all costs duplicating elements, tables, and columns, said integration must also comply with the referential integration, any new component of the application dictionary must be identified with an entity type, the elements, tables, columns, windows, tabs, and fields must be in English each with a fully documented name, description, and help.

6.- The new elements of the application dictionary must have their corresponding XML Script and these must have been generated by reserving ids from the central server, to avoid future collisions with new contributions, if the XML Script includes new database views data these must implement for Oracle and PostgreSQL

7.- The new functionality must be reviewed and approved by the technical team of the project, once all the observations have been resolved, the functional team can be requested to carry out its review and validation process.


Important recommendations . before a sponsor goes launches the development of new functionality, it is very important to request help from the technical and functional teams to discuss their future contribution and, most importantly, its design, this will facilitate integration since there is a design of the data model and functionality was validated previously.


Maybe @marcalwestf Mario and @yamelsenih Yamel have additional suggestions but these are the most important ones that I apply for any new functionality or improvement.

kind regards
Victor

**Comment by e-Evolution** [comment]
Hello everyone!

As the main maintainer of the ADempiere project and a member of the project's functional team, I describe here the main requirements necessary for the integration of any contribution:

Functional Requirements to Accept a Contribution

1.- The improvement that is being promoted must be fully integrated with the rest of the system, if it handles materials with inventory, if it requires planning with MRP or Replenishment, as well as if it requires any monetary management with general accounting, costs, accounts payable or receivable, the integration of the functionality must be general and must be integrated completely and transparently.

The integration must comply with the best practices of behavior patterns that currently exist in the system, such as documents which contain accounting dates, document status, document action, a workflow and their respective functionality and rules for accounting posting, the functionality should not duplicate existing or similar, for example if there is already a product catalog there should not be another model to insert other products, in this case it is recommended to extend the model in another table and add a tab to the current window as per example the functionality of costs, another example is if there is already a way to define work centers or resources, you must use the current functionality, if there is a way to define workflows, routes or quality or maintenance procedures, you must use the functionality of workflows existing, for that they were created, if it requires moving materials, it must be integrated correctly with the  functionality of  material management functionality then need implement the costing and accounting, if there is already a way to perform a job where a repair or maintenance process is required, you must use the current functionality in this case light production or manufacturing orders, we no want other  Maintenance Jobs window when should solve with  a document type, current functionality already handles product inputs and outputs, costs and accounting, work in process, integration with supply chain, purchasing,  project functionality and request management


2.- All the new functionality must be written in English, with names, descriptions in English, and help in English, if there is a translation to another language it is ideal but English is mandatory.

3.- The new functionality must not duplicate the existing functionality, if it requires something that it does not do, it is better to propose an improvement, this often happens due to the sponsor's ignorance of the current functionality of the system.

My recommendation is better to ask before creating a new functionality duplicate so that if a functionality solves a problem that was solved in another way, even without creating new functionality 


4.- Any functionality must be proposite general, so the implementation is also a general solution, a customer's customization should not be confused with new functionality, if the functionality is correctly integrated and complete with the system's functionalities, it will be easy to determine if it is a new functionality that may be included as standard functionality in ADempiere or if just solves a customer specific requirement.

5.- The sponsor must provide the documentation in English to test and validate the functionality. This is necessary so that the functional team of the project can review and will test the functionality.

6.- The sponsor that integrates a new functionality that was accepted commits to maintain it as soon as it is incorporated as standard functionality, this means that it must be responsible for testing, maintaining, and making future bugfix

7.- The sponsor must implement the suggestions and changes requested by the technical or functional teams, they must also be willing to put the necessary resources to complete the integration, you not should think that is  the job of the technical and functional team to complete the integration, we can help with any question, but you need to do the work

One of the frequent problems of sponsors who want to promote some functionality is that they do not meet these technical and functional requirements, which makes the integration not possible, since it does not meet the necessary quality for its integration.

To have a healthy ADempiere it is necessary to meet the quality requirements of the project.

If any of the functional team considers any other requirement that has been omitted, suggestions are accepted.

kind regards
Victor

**Comment by marcalwestf** [comment]
Hi @rvergara Ramiro, thank you for your initiative to integrate Maintenance into ADempiere. I will gladly join your efforts.

In the scrum meeting from yesterday, it was accorded to review the issues @e-Evolution mentions so that they can be valid for any contribution.

**Comment by eduardolopezuy** [comment]
Thank you @rvergara for offering this contribution. We support and we will complement this functionality with some features that we have made. We are waiting to schedule a meeting to work on it.
